mzLib/FlashLFQ/FlashLfqEngine.cs

```diff
@@ -1760,45 +1758,64 @@ public List<IndexedMassSpectralPeak> Peakfind(double idRetentionTime, double mas
                 }
             }
 
+            var xic = GetXIC(mass.ToMz(charge), precursorScanIndex, _peakIndexingEngine, ms1Scans.Length, tolerance, MissedScansAllowed);
+
+            return xic;
+        }
+
+        public static List<IndexedMassSpectralPeak> GetXIC(double mz, int zeroBasedScanIndex, PeakIndexingEngine peakIndexingEngine, int scansLength, Tolerance tolerance, int missedScansAllowed, double maxRT = double.MaxValue)
+        {
+            var xic = new List<IndexedMassSpectralPeak>();
+
             // go right
             int missedScans = 0;
-            for (int t = precursorScanIndex; t < ms1Scans.Length; t++)
+            for (int t = zeroBasedScanIndex; t < scansLength; t++)
             {
-                var peak = _peakIndexingEngine.GetIndexedPeak(mass, t, tolerance, charge);
+                var peak = peakIndexingEngine.GetIndexedPeak(mz, t, tolerance);
 
-                if (peak == null && t != precursorScanIndex)
+                if (peak == null && t != zeroBasedScanIndex)
                 {
                     missedScans++;
                 }
                 else if (peak != null)
                 {
                     missedScans = 0;
                     xic.Add(peak);
+
+                    if (peak.RetentionTime - xic.First().RetentionTime > maxRT)
+                    {
+                        break;
+                    }
                 }
 
-                if (missedScans > MissedScansAllowed)
+                if (missedScans > missedScansAllowed)
                 {
                     break;
                 }
             }
 
             // go left
             missedScans = 0;
-            for (int t = precursorScanIndex - 1; t >= 0; t--)
+            for (int t = zeroBasedScanIndex - 1; t >= 0; t--)
             {
-                var peak = _peakIndexingEngine.GetIndexedPeak(mass, t, tolerance, charge);
+                var peak = peakIndexingEngine.GetIndexedPeak(mz, t, tolerance);
 
-                if (peak == null && t != precursorScanIndex)
+                if (peak == null && t != zeroBasedScanIndex)
                 {
                     missedScans++;
                 }
                 else if (peak != null)
                 {
                     missedScans = 0;
                     xic.Add(peak);
+
+                    if (xic.First().RetentionTime - peak.RetentionTime > maxRT)
+                    {
```

Comment:
Is maxRT a limit to the distance from the point we start looking or maximum retention time? If the first, use a more informative name. If the second, this case is impossible.

Reply:
I think the second case is intended, and it is possible here. In this second loop, we're iterating backwards in time. So the clause subtracts RT of the current point from the RT of the middle point of the XIC